account_loan/report/account_loan_report.py

```diff
@@ -0,0 +1,71 @@
+# Copyright 2018 Ahmed Yousif
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import api, fields, models, tools
+
+class AccountLoanReport(models.Model):
+    _name = "account.loan.report"
+    _description = "Loans Analysis"
+    _auto = False
+
+    name = fields.Char(string='Loan', required=False, readonly=True)
+    description = fields.Char(string='Description', readonly=True)
+    partner_id = fields.Many2one('res.partner', string='Lender', readonly=True)
+    company_id = fields.Many2one('res.company', string='Company', readonly=True)
+    periods = fields.Integer(string='# of Loan Periods', readonly=True)
+    method_period = fields.Integer(string='Time Between Loan Periods', readonly=True)
+    rate = fields.Float(string='Loan Rate', readonly=True)
+    pending_principal_amount = fields.Float(string='Amount of Pending Principal', readonly=True)
+    payment_amount = fields.Float(string='Amount of Payment', readonly=True)
+    interests_amount = fields.Float(string='Amount of Interests', readonly=True)
+    state = fields.Selection([('draft', 'Draft'), ('posted', 'Posted'), ('cancelled', 'Cancelled'), ('close', 'Close')], string='Status', readonly=True)
+    loan_type = fields.Selection([('fixed-annuity', 'Fixed Annuity'), ('fixed-principal', 'Fixed Principal'), ('interest', 'Only interest')], readonly=True)
+    start_date = fields.Date(string='Loan Date', readonly=True)
+    rate_type = fields.Selection([('napr', 'Nominal APR'), ('ear', 'EAR'), ('real', 'Real rate')], readonly=True)
+    loan_amount = fields.Float(string='Loan Amount', readonly=True)
+    short_term_loan_account_id = fields.Many2one('account.account', string='Short term account', readonly=True)
+    long_term_loan_account_id = fields.Many2one('account.account', string='Long term account', readonly=True)
+    interest_expenses_account_id = fields.Many2one('account.account', string='Interests account', readonly=True)
+
+    @api.model_cr
+    def init(self):
```

Comment:
Better use:
    @api.model_cr
    def init(self):
        tools.drop_view_if_exists(self.env.cr, self._table)
        self.env.cr.execute("""CREATE or REPLACE VIEW %s as (
            %s
            FROM ( %s )
            %s
            )""" % (self._table, self._select(), self._from(), self._group_by()))

See for example https://github.com/odoo/odoo/blob/10.0/addons/project/report/project_report.py#L89

Reply:
Thanks again @jbeficent but I don't think we should split one function (method) into three when it's not need and it's sql statement I thik its better stay this way, but I Can make these changes if you insist or we have an issue with this implementation.

Reply:
@imahmedyousif You do it for extensibility purposes. It's a best practice in Odoo. That way one can create modules like account_loan_xxxx that adds fields to the loan and can easily extend the report by calling super to the _select(), _from(),.. methods. Otherwise the sql code is mononlythic and imposible to extend.

Reply:
Thanks for explaining, I have made requested  the changes.